thirdparty/faiss/faiss/utils/onednn_utils.h

```diff
@@ -0,0 +1,216 @@
+ * @param xcol Column number of input matrix X
+ * @param yrow Row number of weight matrix Y
+ * @param ycol Column number of weight matrix Y
+ * @param in_f32_1 Input matrix pointer in float32 type
+ * @param in_f32_2 Weight matrix pointer in float32 type
+ * @param out_f32 Output matrix pointer for result in float32 type
+ * @return None
+ */
+
+enum BASE_DATA_STATE {
+    INIT,
+    MODIFIED,
+    PREPARE,
+    READY
+};
+
+class BaseData {
+ public:
+    static std::atomic<BASE_DATA_STATE>& getState() {
+        static std::atomic<BASE_DATA_STATE> is_base_changed(BASE_DATA_STATE::INIT);
+        return is_base_changed;
+    }
+};
+
+static dnnl::memory bf16_mem2;
+static pthread_rwlock_t rwlock = PTHREAD_RWLOCK_INITIALIZER;
+
+struct inner_product_desc {
+    uint32_t xrow;
+    uint32_t xcol;
+    uint32_t yrow;
+    uint32_t ycol;
+    float* in_f32_1;
+    float* in_f32_2;
+    float* out_f32;
+    dnnl::memory::desc f32_md1;
+    dnnl::memory::desc f32_md2;
+    dnnl::memory::desc f32_dst_md2;
+    dnnl::memory f32_mem1;
+    dnnl::memory f32_mem2;
+    dnnl::memory f32_dst_mem;
+
+    dnnl::memory::desc bf16_md1;
+    dnnl::memory::desc bf16_md2;
+
+    dnnl::inner_product_forward::primitive_desc inner_product_pd;
+    dnnl::inner_product_forward inner_product_prim;
+
+    dnnl::memory bf16_mem1;
+
+    bool is_same(uint32_t xrow, uint32_t xcol, uint32_t yrow, uint32_t ycol,
+      float* in_f32_1, float* in_f32_2) {
+      return this->xrow == xrow && this->xcol == xcol && this->yrow == yrow && this->ycol == ycol && this->in_f32_2 == in_f32_2;
+    }
+
+    void init(uint32_t xrow, uint32_t xcol, uint32_t yrow, uint32_t ycol,
+		float* in_f32_1, float* in_f32_2) {
+        if (is_same(xrow, xcol, yrow, ycol, in_f32_1, in_f32_2)) {
+            if (this->in_f32_1 != in_f32_1) {
+                this->in_f32_1 = in_f32_1;
+                f32_mem1 = dnnl::memory(f32_md1, cpu_engine, in_f32_1);
+            }
+            return;
+        }
+
+        this->xrow = xrow;
+        this->xcol = xcol;
+        this->yrow = yrow;
+        this->ycol = ycol;
+        this->in_f32_1 = in_f32_1;
+        this->in_f32_2 = in_f32_2;
+
+        f32_md1 = dnnl::memory::desc({xrow, xcol}, dnnl::memory::data_type::f32, dnnl::memory::format_tag::ab);
+        f32_md2 = dnnl::memory::desc({yrow, ycol}, dnnl::memory::data_type::f32, dnnl::memory::format_tag::ab);
+        f32_dst_md2 = dnnl::memory::desc({xrow, yrow}, dnnl::memory::data_type::f32, dnnl::memory::format_tag::ab);
+
+        f32_mem1 = dnnl::memory(f32_md1, cpu_engine, in_f32_1);
+        f32_mem2 = dnnl::memory(f32_md2, cpu_engine, in_f32_2);
+        f32_dst_mem = dnnl::memory(f32_dst_md2, cpu_engine);
+
+        // inner memory bf16
+        bf16_md1 = dnnl::memory::desc({xrow, xcol}, dnnl::memory::data_type::bf16, dnnl::memory::format_tag::any);
+        bf16_md2 = dnnl::memory::desc({yrow, ycol}, dnnl::memory::data_type::bf16, dnnl::memory::format_tag::any);
```

Comment:
Noob Q, why we use bf16 here?

Reply:
Because AMX can native support for bf16/int8 compute, which can significantly improve performance, and we have done the test, it have little impact on accuracy.